File: gui/veronica/src/types/protocol.ts
// Incoming message packets from Rust Daemon to React GUI
export type DaemonMessage =
  | { type: "HEARTBEAT"; payload: { timestamp: number; status: string } }
  | { type: "TELEMETRY"; payload: { cpu_usage: number; ram_used_mb: number; ram_total_mb: number } }
  | { type: "LOG"; payload: { level: string; message: string } };

// Outgoing message packets from React GUI to Rust Daemon
export type ClientMessage =
  | { type: "PING" }
  | { type: "COMMAND"; payload: { action: string } };

export type ConnectionStatus = "CONNECTING" | "ONLINE" | "DISCONNECTED";

File: daemon/src/server/protocol.rs
use serde::{Deserialize, Serialize};

#[derive(Debug, Serialize, Deserialize, Clone)]
#[serde(tag = "type", content = "payload")]
pub enum DaemonMessage {
    #[serde(rename = "HEARTBEAT")]
    Heartbeat { timestamp: u64, status: String },

    #[serde(rename = "TELEMETRY")]
    Telemetry { cpu_usage: f32, ram_used_mb: u64, ram_total_mb: u64 },

    #[serde(rename = "LOG")]
    Log { level: String, message: String },
}

#[derive(Debug, Serialize, Deserialize, Clone)]
#[serde(tag = "type", content = "payload")]
pub enum ClientMessage {
    #[serde(rename = "PING")]
    Ping,

    #[serde(rename = "COMMAND")]
    Command { action: String },
}
